### ISSUE TITLE
Test set used as validation set in callbacks

### ISSUE THREAD

**Issue by patrickzib** [titu1994/LSTM-FCN#7]
it seems like the code is using the test set as a validation set (which is not the problem itself):
https://github.com/titu1994/LSTM-FCN/blob/8fc096b969e94362a047f7d5b3a11f9c5a7486c4/utils/keras_utils.py#L83

However, this becomes a problem as 'monitor' is set to 'val_acc' for both model checkpointing and adjusting the learning rate:

https://github.com/titu1994/LSTM-FCN/blob/8fc096b969e94362a047f7d5b3a11f9c5a7486c4/utils/keras_utils.py#L69
I.e., the best model (model checkpoint) is chosen based on the test set accuracy. 

https://github.com/titu1994/LSTM-FCN/blob/8fc096b969e94362a047f7d5b3a11f9c5a7486c4/utils/keras_utils.py#L70
I.e., the learning rate is tuned using the test set accuracy.

Thus, monitor has to be set to monitor='acc' oder monitor='loss' to avoid any test data leakage.
See [Callbacks](https://keras.io/callbacks/#callback)

**Comment by titu1994** [comment]
Hello Dr. Schäfer,

Thank you for pointing this out. This has now been fixed via commit https://github.com/titu1994/LSTM-FCN/commit/f0776e0e5a74a3c65b9ce2c7ca0813b6f2ab4eee

**Comment by patrickzib** [comment]
Great :)

This issue seems to be present in the MLSTM-FCN code, too, with monitor set to val_acc by default:
```python
def train_model(model:Model, dataset_id, dataset_prefix, dataset_fold_id=None, epochs=50, batch_size=128, val_subset=None,
                cutoff=None, normalize_timeseries=False, learning_rate=1e-3, monitor='val_acc', optimization_mode='max', compile_model=True):
``` 
[Source](https://github.com/titu1994/MLSTM-FCN/blob/65c71ccf87b9c482391dec0530003f50b399181e/utils/keras_utils.py#L106)

**Comment by titu1994** [comment]
Hello Dr. Schäfer,

Thank you once again for your help. This has been fixed via commit https://github.com/titu1994/MLSTM-FCN/commit/231cc2fcda9237ee5864d120017e4f71c81d45ae.

**Comment by forestier** [comment]
Hi,
Could you confirm that the classifier performances reported in your papers were computed with the code using the test set as validation set ?
Thanks,
Germain

**Comment by titu1994** [comment]
Hello Dr. Forestier,

For univariate lstm-fcn, the model was trained on only the training set. The weights of the model were updated only on information from the training dataset. 

The results that we reported were of the best possible model. By best possible model we mean that if we stop training at a specific epoch, we will get the reported test set accuracy.

For our multivariate mlstm-fcn paper (which is under revision), we divided the training set into two parts - a train set and a validation dataset.  Those are the results which we will be publishing.

**Comment by patrickzib** [comment]
Thank you for the responses! We and Germain's group seem to have some problems to reproduce the reported result. Would you run the code with the latest bugfix and report the result you get?
Thank you
Patrick

**Comment by fazlekarim** [comment]
Hi,

Sorry for the late response. I seem to be getting similar results on the few data sets I ran it on. 

I think the issue is setting the right number of epoch. We are reporting the best model result assuming we have the right number of epochs down. 

Would you be able to tell us 3-4 datasets you want us to run? Do you want the weight files of them?

Best Regards,
Fazle

**Comment by charlotte-pel** [comment]
Dear Fazle,

What do you mean by the "right number of epochs"? based on the minimum training loss? the maximum test accuracy? an other metric?

Thank you for your clarification.

Sincerely,
Charlotte.

**Comment by fazlekarim** [comment]
Minimum training loss

**Comment by fazlekarim** [comment]
Hey @patrickzib  and @forestier,

Finally made time to run the code on all the datasets with the fix. The results are shown all the way at the bottom of the page of https://github.com/titu1994/LSTM-FCN

**Comment by forestier** [comment]
Hi @fazlekarim

In your code you have different hyperparameters for each dataset (e.g. batch_size, number of epochs, number of units of your LSTM, etc.).

You indicate in your paper "In the initial phase, the optimal hyperparameters for the model are selected for a given dataset. The model is then trained on the given dataset with these hyperparameter settings."

As your previous code used the test_set for model selection, did you also use the test_set to find the hyperparemeters ?

Why did you remove FCN from your new CD Diagram ?

Best regards,

Germain

**Comment by fazlekarim** [comment]
Hey,

So there are two set of results. For the one with model checkpoint of loss, we used minimum train loss to select the hyper parameters (which was only number of cells for the LSTM block). The batch size and number of epoch was kept constant.

As to why I did not put FCN in the CD. I completely forgot to. I will add that and the other NN models. Thank you for letting me know. I will do it in the next 24 hours. Keep in mind that will change the average rankings.

**Comment by forestier** [comment]
Hi,

I'm not sure I understand.

If we look at the python files for each dataset, we see different batch sizes : 

./Phoneme_model.py:    evaluate_model(model, DATASET_INDEX, dataset_prefix='phenome', **batch_size=128**)
./worms_two_class_model.py:    evaluate_model(model, DATASET_INDEX, dataset_prefix='worms_two_class', **batch_size=16**)
./lighting7_model.py:    evaluate_model(model, DATASET_INDEX, dataset_prefix='lighting7', **batch_size=32**,
(...)

different number of epochs :

./Phoneme_model.py:    #train_model(model, DATASET_INDEX, dataset_prefix='phenome', **epochs=2000**, batch_size=64)
./worms_two_class_model.py:    #train_model(model, DATASET_INDEX, dataset_prefix='worms_two_class', **epochs=1000,** batch_size=16)
./lighting7_model.py:    #train_model(model, DATASET_INDEX, dataset_prefix='lighting7', **epochs=3000**, batch_size=32,
(...)

and different number of units for the LSTM :

./Phoneme_model.py:    x = **LSTM(8)**(ip)
./uwave_gesture_library_y_model.py:    x = **LSTM(128)**(ip)
./wine_model.py:    x = **LSTM(64)**(ip)
(...)

So I'm trying to understand how you chose these parameters for each dataset. Is the code for hyperparameters search available ?

Best regards,

Germain

**Comment by fazlekarim** [comment]
So we re-ran all the experiments. We kept the batch size a constant of 128 and epoch of 2000. The scripts are all old and not updated. We will remove all the individual script files and have one script file that reflects this. 

For the LSTM size what we did was train the model on 8, 64, 128. Then had a separate script (which we will add soon) that selected the model with the minimum training loss. Then we evaluated that model on the test. That is the score you see in the table.

**Comment by forestier** [comment]
Thanks, it's clearer now, looking forward to trying your code.

Best regards,

Germain

**Comment by fazlekarim** [comment]
Thanks. I really appreciate all the feedback. I will let you know when it's up.

**Comment by patrickzib** [comment]
Hi @fazlekarim, 

thank you very much for your update! I do not totally understand the accuracy tables listed in the web page. The second table  refers to the fix and the critical difference diagram still refers to the published results without the fix? Correct? 

Also I do not understand the implications of the dimension shuffle prior to the LSTM block. I understand that it is beneficial for accuracy (as you have shown) but in concept isn't this more like a fully connected network with n dimensions? At least it is not a recurrent network anymore as you cannot input variable length data  (this would result in variable length dimensionality after dimension shuffling and you would have to do padding to fix dimensionality, also the offset of a feature matters after dimension shuffle?), or am I seeing this wrong? 

Just a guess: Have you ever tried using a fully connected network as a replacement of this lstm block? 

Best

**Comment by fazlekarim** [comment]
Hi @patrickzib ,

The second table, critical difference diagram, and P-value table refer to the fix. This is based on the minimum training loss. I uploaded my results as a csv file in the Results Folder, in case I made a mistake computing the metrics.

In regards to your second question, yes it is more like a fully connected network with n dimensions. You are seeing this correctly. The only difference is the computation happening in the cell when compared to the fully connected network. 

In fact, we have. We tried replacing it with a GRU, RNN, Fully connected network. We also applied LSTM without dimension shuffle. For some reason, LSTM-FCN with dimension shuffle outperforms everything (I am not going to lie, the results suprised me too). We have all the results in this repo: https://github.com/titu1994/LSTM-FCN-ablation. I made a mistake in the P-Value table of the paper associated with the repo. So I updated the arxiv paper yesterday with the fix. It will take a few days for Arxiv to upload the updated paper. If you want the fixed paper, I can email it to you.

**Comment by titu1994** [comment]
@patrickzib Hello Dr. Schäfer,

Regarding your second question, I would like to extend Fazle's reply. 

After a dimension shuffle, the neural network as a whole can no longer accept timeseries possessing dynamic timesteps. The reason for this is that when the model is being built by Keras / Tensorflow, it must determine the input dimension to construct the weight matrices of the subsequent layer.

As the weight matrices are constructed just once (based on the input shape provided the first time the model is built), dynamic number of timesteps is no longer possible as after dimension shuffle, the number of timesteps `T` becomes the input dimension of the LSTM. The same concepts apply to PyTorch as well, as there we must specify the input dimensions when defining the layers of the model itself.

If dynamic timesteps are absolutely necessary for some task, then as you suggested, one would need to either pad with zeros or clip at the end. This may or may not harm the performance of the model. 

Finally, as to whether after dimension shuffle, the lstm pathway is still recurrent. In the univariate case, you are absolutely correct in your conclusion. 

However, when the original input is a multivariate time series (characterized as a tensor of shape (N, T, V)), where N is the number of samples in the dataset, T is the number of timesteps and V as the number of correlated variables per timestep in the sequence, then after the dimension shuffle, recurrent behavior occurs again but on variables rather than timesteps. This is also the primary reason why we do no simply convert the LSTM cell into an equivalent gated Dense layer - to enable easy extension to the multivariate case.
